File: tiq-e/strings/9/main.go
package main

import (
	"fmt"
	"time"
)

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	var (
		lenPrefix int
		curChar   byte
	)

	for {
		if lenPrefix >= len(strs[0]) {
			break
		}

		curChar = strs[0][lenPrefix]

		for _, w := range strs {
			if lenPrefix >= len(w) || curChar != w[lenPrefix] {
				return w[:lenPrefix]
			}
		}

		lenPrefix += 1
	}
	return strs[0][:lenPrefix]
}

func main() {
	var arr = []string{"flower", "flow", "flight"}

	var start = time.Now()
	var result = longestCommonPrefix(arr)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}

File: tiq-e/strings/9/main.rs
use std::time::Instant;

struct Solution;

impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        if strs.len() == 0 {
            return String::new();
        }
        let mut template = strs[0].to_string();

        while template.len() > 0 {
            let mut matched_all = true;
            for s in &strs {
                if !s.starts_with(&template) {
                    matched_all = false;
                    break;
                }
            }
            if matched_all {
                return template;
            } else {
                template.pop();
            }
        }
        template
    }
}

fn main() {
    let arr = [String::from("flower"), String::from("flow"), String::from("flight")];

    let start = Instant::now();
    let result = Solution::longest_common_prefix(Vec::from(arr));
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
